In [66]:
import pandas as pd
import numpy as np
from sklearn.metrics import confusion_matrix

### Model Evaluation Exercise

#### (1) Create a new file named model_evaluation.py or model_evaluation.ipynb for these exercises.

##### (2) Given the following confusion matrix, evaluate (by hand) the model's performance.


|               | actual cat | actual dog |
|:------------  |-----------:|-----------:|
| predicted cat |         34 |          7 |
| predicted dog |         13 |         46 |


- In the context of this problem, what is a false positive?
    - <b><i>Predicting cat when actually dog</i></b>
- In the context of this problem, what is a false negative?
    - <b><i>Predicting dog when actually cat</i></b>
- How would you describe this model?
    - <b><i>When looking at the model through the lense of 'actual cat' being our True Positive, the accuracy is 89% overall, the precision show we predicted people like cats right 83% of the time and our recall showed we predicted 72% of the time that people liked cats

#### (3) You are working as a datascientist working for Codeup Cody Creator (C3 for short), a rubber-duck manufacturing plant.

- Unfortunately, some of the rubber ducks that are produced will have defects. Your team has built several models that try to predict those defects, and the data from their predictions can be found here.

Use the predictions dataset and pandas to help answer the following questions:

In [54]:
c3 = pd.read_csv(r'/Users/davidberchelmann/codeup-data-science/classification-exercises/cs3.csv')

In [55]:
c3.head()

actual     model1  model2     model3
0  No Defect  No Defect  Defect  No Defect
1  No Defect  No Defect  Defect     Defect
2  No Defect  No Defect  Defect  No Defect
3  No Defect     Defect  Defect     Defect
4  No Defect  No Defect  Defect  No Defect

##### - An internal team wants to investigate the cause of the manufacturing defects. They tell you that they want to identify as many of the ducks that have a defect as possible. Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

In [56]:
pd.crosstab(c3.actual, c3.model1)

model1     Defect  No Defect
actual                      
Defect          8          8
No Defect       2        182

In [57]:
pd.crosstab(c3.actual, c3.model2)

model2     Defect  No Defect
actual                      
Defect          9          7
No Defect      81        103

In [87]:
pd.crosstab(c3.actual, c3.model3)

model3     Defect  No Defect
actual                      
Defect         13          3
No Defect      86         98

In [59]:
c3.actual.value_counts()

No Defect    184
Defect        16
Name: actual, dtype: int64

In [85]:
c3.model3.value_counts()

No Defect    101
Defect        99
Name: model3, dtype: int64

In [61]:
c3['baseline_prediction'] = 'Defect'

In [62]:
model_accuracy = (c3.model3 == c3.actual).mean()
baseline_accuracy = (c3.baseline_prediction == c3.actual).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 55.50%
baseline accuracy: 8.00%


In [63]:
y_trues = c3.actual
y_pred = c3.model1


In [64]:
(tp, fn, fp, tn) = confusion_matrix(y_trues, y_pred,
                                    labels = ['Defect', 'No Defect']).ravel()
(tp, fn, fp, tn)

(8, 8, 2, 182)

In [43]:
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print("Accuracy is", accuracy)
print("Recall is", recall)
print("Precision is", precision)

True Positives 8
False Positives 2
False Negatives 8
True Negatives 182
-------------
Accuracy is 0.95
Recall is 0.5
Precision is 0.8


In [44]:
c3['baseline_prediction'] = 'Defect'

In [45]:
(c3.actual == c3.baseline_prediction).mean()

0.08

In [46]:
model_accuracy = (c3.actual == c3.model3).mean()
baseline_accuracy = (c3.actual == c3.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 55.50%
baseline accuracy: 8.00%


In [47]:
subset = c3[c3.actual == 'Defect']
subset

actual     model1     model2     model3 baseline_prediction
13   Defect  No Defect     Defect     Defect              Defect
30   Defect     Defect  No Defect     Defect              Defect
65   Defect     Defect     Defect     Defect              Defect
70   Defect     Defect     Defect     Defect              Defect
74   Defect  No Defect  No Defect     Defect              Defect
87   Defect  No Defect     Defect     Defect              Defect
118  Defect  No Defect     Defect  No Defect              Defect
135  Defect     Defect  No Defect     Defect              Defect
140  Defect  No Defect     Defect     Defect              Defect
147  Defect     Defect  No Defect     Defect              Defect
163  Defect     Defect     Defect     Defect              Defect
171  Defect  No Defect     Defect     Defect              Defect
176  Defect  No Defect     Defect     Defect              Defect
186  Defect  No Defect  No Defect  No Defect              Defect
194  Defect     Defect  No Defect     Defect              Defect
196  Defect     Defect  No Defect  No Defect              Defect

In [48]:
model_recall = (subset.actual == subset.model3).mean()
baseline_recall = (subset.actual == subset.baseline_prediction).mean()

print(f'   model recall: {model_recall:.2%}')
print(f'baseline recall: {baseline_recall:.2%}')

   model recall: 81.25%
baseline recall: 100.00%


In [49]:
subset = c3[c3.model3 == 'Defect']
subset

actual     model1     model2  model3 baseline_prediction
1    No Defect  No Defect     Defect  Defect              Defect
3    No Defect     Defect     Defect  Defect              Defect
5    No Defect  No Defect  No Defect  Defect              Defect
9    No Defect  No Defect  No Defect  Defect              Defect
13      Defect  No Defect     Defect  Defect              Defect
..         ...        ...        ...     ...                 ...
193  No Defect  No Defect     Defect  Defect              Defect
194     Defect     Defect  No Defect  Defect              Defect
195  No Defect  No Defect     Defect  Defect              Defect
198  No Defect  No Defect     Defect  Defect              Defect
199  No Defect  No Defect  No Defect  Defect              Defect

[99 rows x 5 columns]

In [50]:
model_precision = (subset.actual == subset.model3).mean()
print(f'model precision: {model_precision:.2%}')

model precision: 13.13%


In [51]:
subset = c3[c3.baseline_prediction == 'Defect']
baseline_precision = (subset.actual == subset.model3).mean()

print(f'baseline precision: {baseline_precision:.2%}')

baseline precision: 55.50%


- <span style="color:red"><b> - The best model to use will be model 3 since it was able to identify the most defects and best metric is recall since it does the best job of capturing Defects.</b></span>

###### Recently several stories in the local news have come out highlighting customers who received a rubber duck with a defect, and portraying C3 in a bad light. 

The PR team has decided to launch a program that gives customers with a defective duck a vacation to Hawaii. They need you to predict which ducks will have defects, but tell you the really don't want to accidentally give out a vacation package when the duck really doesn't have a defect. 

Which evaluation metric would be appropriate here? Which model would be the best fit for this use case?

- <font color=red><b>Precision is the best metric to use here since false positives are low while model 1 will be the best fit to evaluate since it has the highest precision rate out of the 3.</b></font>

#### You are working as a data scientist for Gives You Paws ™, a subscription based service that shows you cute pictures of dogs or cats (or both for an additional fee).

At Gives You Paws, anyone can upload pictures of their cats or dogs. The photos are then put through a two step process. First an automated algorithm tags pictures as either a cat or a dog (Phase I). Next, the photos that have been initially identified are put through another round of review, possibly with some human oversight, before being presented to the users (Phase II).

Given this dataset, use pandas to create a baseline model (i.e. a model that just predicts the most common class) and answer the following questions:

- <font color = blue>In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?</font>



- <font color = green> Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?</font>



- <font color = purple>Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?</font>


In [52]:
 paws = pd.read_csv(r'/Users/davidberchelmann/codeup-data-science/classification-exercises/gives_you_paws.csv')

In [29]:
paws.head()

actual model1 model2 model3 model4
0    cat    cat    dog    cat    dog
1    dog    dog    cat    cat    dog
2    dog    cat    cat    cat    dog
3    dog    dog    dog    cat    dog
4    cat    cat    cat    dog    dog

In [91]:
paws['actual'].value_counts()

dog    3254
cat    1746
Name: actual, dtype: int64

In [92]:
paws['baseline_prediction'] = 'dog'

In [93]:
(paws.actual == paws.baseline_prediction).mean()

0.6508

In [94]:
paws.head()

actual model1 model2 model3 model4 baseline_prediction
0    cat    cat    dog    cat    dog                 dog
1    dog    dog    cat    cat    dog                 dog
2    dog    cat    cat    cat    dog                 dog
3    dog    dog    dog    cat    dog                 dog
4    cat    cat    cat    dog    dog                 dog

In [95]:
pd.crosstab(paws.actual, paws.model1)

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [114]:
y_true = paws.actual
y_pred = paws.model2


In [115]:
(tn, fp, fn, tp) = confusion_matrix(y_true, y_pred,
                                    labels = ['cat', 'dog']).ravel()
(tn, fp, fn, tp)

(1555, 191, 1657, 1597)

In [116]:
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print("Accuracy is", accuracy)
print("Recall is", recall)
print("Precision is", precision)

True Positives 1597
False Positives 191
False Negatives 1657
True Negatives 1555
-------------
Accuracy is 0.6304
Recall is 0.49078057775046097
Precision is 0.8931767337807607


In [99]:
models = list(paws.columns)
models = models[1:]
models

['model1', 'model2', 'model3', 'model4', 'baseline_prediction']

In [100]:
output = {}
for model in models:
    accuracy = (paws.actual == paws[model]).mean()
    output.update({model:accuracy})
    
    
output

{'model1': 0.8074,
 'model2': 0.6304,
 'model3': 0.5096,
 'model4': 0.7426,
 'baseline_prediction': 0.6508}

In [102]:
pd.DataFrame(list(output.items()), columns = ['model', 'accuracy'])

model  accuracy
0               model1    0.8074
1               model2    0.6304
2               model3    0.5096
3               model4    0.7426
4  baseline_prediction    0.6508

In [101]:
model_accuracy = (paws.actual == paws.model1).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 80.74%
baseline accuracy: 65.08%


- <font color = blue>In terms of accuracy, how do the various models compare to the baseline model? Are any of the models better than the baseline?</font>
    - <b>Models 1 & 4 beat the baseline in terms of accuracy. Baseline accuracy is 65.08%. Model 1 cames in at 80.74% while Model 4 is at 74.26%

- <font color = green> Suppose you are working on a team that solely deals with dog pictures. Which of these models would you recomend for Phase I? For Phase II?</font>
    - <b> For phase 1 where precision is important, I would go with Model 1. When showing to users where it's important to make sure positive dog photos are captured so I would go with Model 4 using recall

- my notes.....Model 4 has a high recal
in the 90s, Model 2 is best overall with highest precision at 89...Model 1 is close as well in regards to precision...89 as well. Model 3 is poor all around

- correct answer for phase 1 is recall and then precision for phase 2 (phase 2, we want to minimize false positives)

In [33]:
 pawsc = pd.read_csv(r'/Users/davidberchelmann/codeup-data-science/classification-exercises/gives_you_paws.csv')

In [34]:
pawsc['baseline_prediction'] = 'cat'

In [77]:
y_true_cat = pawsc.actual
y_pred_cat = pawsc.model2

In [78]:
pd.crosstab(pawsc.actual, pawsc.model1)

model1   cat   dog
actual            
cat     1423   323
dog      640  2614

In [79]:
(tn, fp, fn, tp) = confusion_matrix(y_true_cat, y_pred_cat,
                                    labels = ['dog', 'cat']).ravel()
(tn, fp, fn, tp)

(1597, 1657, 191, 1555)

In [80]:
print("True Positives", tp)
print("False Positives", fp)
print("False Negatives", fn)
print("True Negatives", tn)

print("-------------")

accuracy = (tp + tn) / (tp + tn + fp + fn)
recall = tp / (tp + fn)
precision = tp / (tp + fp)

print("Accuracy is", accuracy)
print("Recall is", recall)
print("Precision is", precision)

True Positives 1555
False Positives 1657
False Negatives 191
True Negatives 1597
-------------
Accuracy is 0.6304
Recall is 0.8906071019473081
Precision is 0.4841220423412204


In [81]:
model_accuracy = (paws.actual == paws.model1).mean()
baseline_accuracy = (paws.actual == paws.baseline_prediction).mean()

print(f'   model accuracy: {model_accuracy:.2%}')
print(f'baseline accuracy: {baseline_accuracy:.2%}')

   model accuracy: 80.74%
baseline accuracy: 65.08%


- <font color = purple>Suppose you are working on a team that solely deals with cat pictures. Which of these models would you recomend for Phase I? For Phase II?</font>

    - For phase 1, model 4 has the best precision at over 80%. For phase II, I'd go with model 2 which has a recall of 89%

In [68]:
# import sklearn functions
from sklearn.metrics import precision_score
from sklearn.metrics import accuracy_score
from sklearn.metrics import recall_score
from sklearn.metrics import classification_report

### Metrics for dogs & cats

In [69]:
# the below is just for dogs, cats would be the same, just pawsc. Great thing is that this can all be accomplished by 
# doing a classification report

y_true = paws.actual
y_mod1 = paws.model1
y_mod2 = paws.model2
y_mod3 = paws.model3
y_mod4 = paws.model4

In [70]:
accuracy_score(y_true, y_mod1)

0.8074

In [71]:
recall_score(y_true, y_mod1,
            labels = ['cat', 'dog'],
            average = None)

array([0.81500573, 0.80331899])

In [72]:
precision_score(y_true, y_mod1,
            labels = ['cat', 'dog'],
            average = None)

array([0.68977218, 0.89002383])

In [122]:
# classification report for dogs/cats
print(classification_report(y_true, y_mod1, labels = ['cat', 'dog']))


              precision    recall  f1-score   support

         cat       0.69      0.82      0.75      1746
         dog       0.89      0.80      0.84      3254

    accuracy                           0.81      5000
   macro avg       0.79      0.81      0.80      5000
weighted avg       0.82      0.81      0.81      5000



In [76]:
# classification report for defects/no defects (model 1)
print(classification_report(y_trues, y_pred, labels = ['Defect', 'No Defect']))

              precision    recall  f1-score   support

      Defect       0.80      0.50      0.62        16
   No Defect       0.96      0.99      0.97       184

    accuracy                           0.95       200
   macro avg       0.88      0.74      0.79       200
weighted avg       0.95      0.95      0.94       200

